### PR TITLE
Coinmate stream WebSocket API 2.0

### PR DESCRIPTION
**Pull request by stachon**
Migate stream-coinmate to use WebSocket API 2.0, which uses channel subscriptions over one connection instead of one connection per channel.

### CODE REVIEW

**Comment by timmolter** [comment]
Please fix the test failures.

File: xchange-stream-coinmate/src/main/java/info/bitrich/xchangestream/coinmate/v2/CoinmateStreamingService.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+  CoinmateStreamingService(String url, AuthParams authParams) {
+    super(url);
+    this.authParams = authParams;
+  }
+
+  @Override
+  protected String getChannelNameFromMessage(JsonNode message) throws IOException {
+    String event = message.get("event").asText();
+    if (!event.equals("data")) {
```
Comment (earce) [inline xchange-stream-coinmate/src/main/java/info/bitrich/xchangestream/coinmate/v2/CoinmateStreamingService.java:31]: put constant on left side to protect NPE

File: xchange-stream-coinmate/src/main/java/info/bitrich/xchangestream/coinmate/v2/CoinmateStreamingService.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    return message.get("channel").asText();
+  }
+
+  @Override
+  public String getSubscribeMessage(String channelName, Object... args) throws IOException {
+    if (args.length > 0 && args[0].equals(true)) {
+      // needs authentication
+      CoinmateAuthenticedSubscribeMessage subscribeMessage =
+          generateAuthenticatedSubscribeMessage(channelName);
+      return objectMapper.writeValueAsString(subscribeMessage);
```
Comment (earce) [inline xchange-stream-coinmate/src/main/java/info/bitrich/xchangestream/coinmate/v2/CoinmateStreamingService.java:43]: can you just inline this here and below as opposed to creating variables

File: xchange-stream-coinmate/src/main/java/info/bitrich/xchangestream/coinmate/v2/CoinmateStreamingService.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+          generateSubscribeMessage(channelName);
+      return objectMapper.writeValueAsString(subscribeMessage);
+    }
+  }
+
+  @Override
+  public String getUnsubscribeMessage(String channelName) throws IOException {
+    CoinmateUnsubscribeMessage unsubscribeMessage =
+        generateUnsubscribeMessage(channelName);
+    return objectMapper.writeValueAsString(unsubscribeMessage);
```
Comment (earce) [inline xchange-stream-coinmate/src/main/java/info/bitrich/xchangestream/coinmate/v2/CoinmateStreamingService.java:55]: inline

File: xchange-stream-coinmate/src/main/java/info/bitrich/xchangestream/coinmate/v2/CoinmateStreamingService.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+  @Override
+  protected void handleIdle(ChannelHandlerContext ctx) {
+    // the default zero frame is not handled by Coinmate API, use ping message instead
+    String pingMessage = null;
+    try {
+      pingMessage = objectMapper.writeValueAsString(new CoinmatePingMessage());
+    } catch (JsonProcessingException e) {
+    }
+    ctx.writeAndFlush(new TextWebSocketFrame(pingMessage));
```
Comment (earce) [inline xchange-stream-coinmate/src/main/java/info/bitrich/xchangestream/coinmate/v2/CoinmateStreamingService.java:66]: remove variable, put writeAndFlush inside try/catch and add log line,

File: xchange-stream-coinmate/src/main/java/info/bitrich/xchangestream/coinmate/v2/CoinmateStreamingService.java (hunk 1 above)
Comment (stachon) [reply xchange-stream-coinmate/src/main/java/info/bitrich/xchangestream/coinmate/v2/CoinmateStreamingService.java:31]: Thanks

File: xchange-stream-coinmate/src/main/java/info/bitrich/xchangestream/coinmate/v2/CoinmateStreamingService.java (hunk 2 above)
Comment (stachon) [reply xchange-stream-coinmate/src/main/java/info/bitrich/xchangestream/coinmate/v2/CoinmateStreamingService.java:43]: Please, what do you mean by inlining? Constructing the JSON manually as a String?

Comment (earce) [reply xchange-stream-coinmate/src/main/java/info/bitrich/xchangestream/coinmate/v2/CoinmateStreamingService.java:43]:
```
      return objectMapper.writeValueAsString(
          generateAuthenticatedSubscribeMessage(channelName));
```

File: xchange-stream-coinmate/src/main/java/info/bitrich/xchangestream/coinmate/v2/CoinmateStreamingService.java (hunk 4 above)
Comment (stachon) [reply xchange-stream-coinmate/src/main/java/info/bitrich/xchangestream/coinmate/v2/CoinmateStreamingService.java]: thanks, fixed

File: xchange-stream-coinmate/src/main/java/info/bitrich/xchangestream/coinmate/v2/CoinmateStreamingService.java (hunk 3 above)
Comment (stachon) [reply xchange-stream-coinmate/src/main/java/info/bitrich/xchangestream/coinmate/v2/CoinmateStreamingService.java]: thanks, fixed

File: xchange-stream-coinmate/src/main/java/info/bitrich/xchangestream/coinmate/v2/CoinmateStreamingService.java (hunk 2 above)
Comment (stachon) [reply xchange-stream-coinmate/src/main/java/info/bitrich/xchangestream/coinmate/v2/CoinmateStreamingService.java]: thanks, fixed

File: xchange-stream-coinmate/src/test/java/info/bitrich/xchangestream/coinmate/v2/CoinmateManualExample.java [hunk 5]
```diff
@@ -11,17 +15,26 @@
...
     exchange.connect().blockingAwait();
 
-    exchange
-        .getStreamingMarketDataService()
-        .getOrderBook(CurrencyPair.BTC_EUR)
-        .subscribe(
-            orderBook -> {
-              LOG.info("Ask: {}", orderBook.getAsks().get(0));
-              LOG.info("Bid: {}", orderBook.getBids().get(0));
+//    exchange
```
Comment (earce) [inline xchange-stream-coinmate/src/test/java/info/bitrich/xchangestream/coinmate/v2/CoinmateManualExample.java:26]: can you get rid of this comment?

Comment (stachon) [reply xchange-stream-coinmate/src/test/java/info/bitrich/xchangestream/coinmate/v2/CoinmateManualExample.java]: Thanks, also removed credentials

Comment (stachon) [reply xchange-stream-coinmate/src/test/java/info/bitrich/xchangestream/coinmate/v2/CoinmateManualExample.java]: Thanks, also removed credentials

Comment (earce) [reply xchange-stream-coinmate/src/test/java/info/bitrich/xchangestream/coinmate/v2/CoinmateManualExample.java]: You should delete those api keys in your app so as to not compromise yourself

Comment (stachon) [reply xchange-stream-coinmate/src/test/java/info/bitrich/xchangestream/coinmate/v2/CoinmateManualExample.java]: Thanks for noticing! I have disabled those API keys

**Pull request merged**